ISPyB LIMS: add a hook method for ISPyB authentication

elmjag opened the pull request:
Add an abstract method `ispyb_login()`, which is invoked when auth server is set to `ispyb`.

This allow to implement ISPyB authentication by overriding this method in your custom ISPyB lims class.

Commit:
ISPyB LIMS: add a hook method for ISPyB authentication

Add an abstract method ispyb_login(), which is invoked when auth
server is set to 'ispyb'.

This allow to implement ISPyB authentication by overriding this
method in your custom ISPyB lims class.

fabcor-maxiv commented on mxcubecore/HardwareObjects/abstract/ISPyBAbstractLims.py:
```diff
@@ -129,6 +129,9 @@ def ldap_login(self, login_name, psd):
         return self.ldapConnection.authenticate(login_name, psd)
 
+    def ispyb_login(self, login_name, psd):
+        raise Exception("Abstract class. Not implemented")
```
```suggestion
        raise NotImplementedError
```

https://docs.python.org/3/library/exceptions.html#NotImplementedError

elmjag replied:
This is the style used in this class, but I agree with @fabcor-maxiv, it should raise `NotImplementedError`. I'll change it everywhere.

Also, it should be:

`raise NotImplementedError()`

😅

fabcor-maxiv replied:
I read somewhere that `raise SomeException` ends up running faster in some cases but never slower than `raise SomeException()`. There is also this linting rule: <https://docs.astral.sh/ruff/rules/unnecessary-paren-on-raise-exception/>.

Anyway, with or without parenthesises, my point is that we should use `NotImplementedError`.

elmjag replied:
Changed to raise `NotImplementedError` everywhere.

Commit:
ISPyB LIMS: raise NotImplementedError

Reise the dedicated 'NotImplementedError' exception if an
'abstract' method is invoked in ISPyBAbstractLIMS class.

This is the established way to deal with 'abstart' method in python.

marcus-oscarsson commented:
Ok, same as the others looks good. You can rebase and merge :)
